# Explore the recombinant origins of B.1.617.2

In [1]:
import numpy as np
import tskit
import tszip
import pandas as pd

import matplotlib.pyplot as plt
import sys

sys.path.append("../")
import sc2ts

In [2]:
ts_path  ="../results/DELETE_v1-alpha4-mds1000/DELETE_v1-alpha4-mds1000_2021-06-21.ts.tsz"
ds_path = "../data/viridian_2024-04-29.alpha1.zarr.zip"
ts = tszip.load(ts_path)
ts

In [3]:
ti = sc2ts.TreeInfo(ts)

def report(*args, **kwargs):
    display(*ti.node_report(*args, **kwargs))



Counting descendants :   0%|          | 0/182131 [00:00<?, ?it/s]

Indexing metadata    :   0%|          | 0/182131 [00:00<?, ?it/s]

Classifying mutations:   0%|          | 0/249569 [00:00<?, ?it/s]

In [4]:
first_617_2 = ti.pango_lineage_samples["B.1.617.2"][0]
focal_strain = ti.nodes_metadata[first_617_2]["strain"]

In [5]:
report(first_617_2)

# Report for 131031

[{'Artic_primer_version': '3',
  'Collection_date': '2021-03-23',
  'Country': 'United Kingdom',
  'Date_tree': '2021-03-23',
  'Date_tree_order': 'Date',
  'Experiment': 'ERX5392273',
  'First_created': '2021-04-06',
  'Genbank_N': 0,
  'Genbank_accession': 'OU214662.1',
  'Genbank_other_runs': '.',
  'Genbank_pangolin': 'B.1.617.2',
  'Genbank_scorpio': 'Delta (B.1.617.2-like)',
  'Genbank_tree_name': 'ERR5676810.genbank.OU214662.1',
  'In_Viridian_tree': True,
  'In_intersection': True,
  'In_may_2024_preprint': True,
  'Platform': 'ILLUMINA',
  'Region': 'none',
  'Run_count': 1,
  'Sample': 'SAMEA8526330',
  'Study': 'PRJEB37886',
  'Viridian_N': 0,
  'Viridian_amplicon_scheme': 'COVID-ARTIC-V3',
  'Viridian_cons_het': 0,
  'Viridian_cons_len': 29823,
  'Viridian_pangolin': 'B.1.617.2',
  'Viridian_pangolin_1.29': 'B.1.617.2',
  'Viridian_result': 'PASS',
  'Viridian_scorpio': 'Delta (B.1.617.2-like)',
  'Viridian_scorpio_1.29': 'Delta (B.1.617.2-like)',
  'date': '2021-03-23',
  

### Node data 

node     flags                               strain      pango  parents  \
0  131031  S______U                           ERR5676810  B.1.617.2        1   
1  135884  __P_____  2021-03-30:A8358T, G17721T, G25494C                   1   

   children  exact_matches  descendants       date  mutations  reversions  \
0         0              1            1 2021-03-23          3           0   
1        60              4          236 2021-03-23          1           0   

   immediate_reversions  child_mutations  child_reversions    role  \
0                     0                0                 0   Child   
1                     0              140                 0  Parent   

   branch_length  
0   0.000000e+00  
1   1.562501e-08

### Edges

### Copying pattern

## Left path to root

### Closest recombinant: 131032

node     flags                                             strain  \
0   131031  S______U                                         ERR5676810   
1   135884  __P_____                2021-03-30:A8358T, G17721T, G25494C   
2   155262  __P_____                                 2021-04-28:C24745T   
3   141988  _O______       2021-04-08:C27739T, C1267T, A20262G, T12946C   
4   134638  __P_____  2021-03-28:C1267T, C6539T, A8358T, T12946C, G1...   
5   142053  __P_____                         2021-04-08:C9891T, T11418C   
6   149557  __P_____                        2021-04-19:G15451A, C16466T   
7   131032  ___R____                                           2c320854   
8     5299  ________                                           dc01abc3   
9      996  S_______                                        SRR11513115   
10      58  _O______                                 2020-02-03:C14408T   
11      15  _O______                  2020-01-28:A23403G, C3037T, C241T   
12      30  _O______                                  2020-01-30:C3037T   
13       1  ______F_                                    Wuhan/Hu-1/2019   

        pango  parents  children  exact_matches  descendants       date  \
0   B.1.617.2        1         0              1            1 2021-03-23   
1                    1        60              4          236 2021-03-23   
2                    1         2              0          262 2021-03-23   
3                    1         3              0          593 2021-03-23   
4                    1         2              0         1804 2021-03-23   
5                    1         2              0        10584 2021-03-23   
6                    1         2              0        10685 2021-03-23   
7                    2         1              0        10685 2021-03-23   
8                    1        14             26        10704 2020-04-02   
9     B.1.384        1        10              5        10722 2020-03-19   
10                   1       302            257       150387 2020-01-19   
11                   1         3              2       150390 2020-01-11   
12                   1         3              0       150395 2020-01-03   
13                   1       130             48       152905 2019-12-26   

    mutations  reversions  immediate_reversions  branch_length  
0           3           0                     0   1.562501e-08  
1           1           0                     0   6.835933e-09  
2           3           0                     0   4.785156e-08  
3           4           0                     0   5.468750e-08  
4           2           0                     0   1.093750e-07  
5           2           0                     0   9.570313e-08  
6           9           0                     0   6.699219e-07  
7           0           0                     0   3.550000e+02  
8           1           0                     0   1.400000e+01  
9           2           0                     0   6.000000e+01  
10          1           0                     0   8.000000e+00  
11          2           0                     0   8.000000e+00  
12          1           0                     0   8.000000e+00  
13          0           0                     0   1.000000e+00

## Right path to root

### Closest recombinant: 131032

node     flags                                             strain  \
0   131031  S______U                                         ERR5676810   
1   135884  __P_____                2021-03-30:A8358T, G17721T, G25494C   
2   155262  __P_____                                 2021-04-28:C24745T   
3   141988  _O______       2021-04-08:C27739T, C1267T, A20262G, T12946C   
4   134638  __P_____  2021-03-28:C1267T, C6539T, A8358T, T12946C, G1...   
5   142053  __P_____                         2021-04-08:C9891T, T11418C   
6   149557  __P_____                        2021-04-19:G15451A, C16466T   
7   131032  ___R____                                           2c320854   
8   118401  __P_____               2021-03-02:C20936T, C24863G, T27299C   
9   113172  _O______                                 2021-02-22:C25276T   
10      58  _O______                                 2020-02-03:C14408T   
11      15  _O______                  2020-01-28:A23403G, C3037T, C241T   
12      30  _O______                                  2020-01-30:C3037T   
13       1  ______F_                                    Wuhan/Hu-1/2019   

        pango  parents  children  exact_matches  descendants       date  \
0   B.1.617.2        1         0              1            1 2021-03-23   
1                    1        60              4          236 2021-03-23   
2                    1         2              0          262 2021-03-23   
3                    1         3              0          593 2021-03-23   
4                    1         2              0         1804 2021-03-23   
5                    1         2              0        10584 2021-03-23   
6                    1         2              0        10685 2021-03-23   
7                    2         1              0        10685 2021-03-23   
8                    1         3              0        10810 2021-01-13   
9                    1         2              0        10811 2020-04-04   
10                   1       302            257       150387 2020-01-19   
11                   1         3              2       150390 2020-01-11   
12                   1         3              0       150395 2020-01-03   
13                   1       130             48       152905 2019-12-26   

    mutations  reversions  immediate_reversions  branch_length  
0           3           0                     0   1.562501e-08  
1           1           0                     0   6.835933e-09  
2           3           0                     0   4.785156e-08  
3           4           0                     0   5.468750e-08  
4           2           0                     0   1.093750e-07  
5           2           0                     0   9.570313e-08  
6           9           0                     0   6.699219e-07  
7           0           0                     0   6.956250e+01  
8          16           0                     0   2.839375e+02  
9           1           0                     0   7.550000e+01  
10          1           0                     0   8.000000e+00  
11          2           0                     0   8.000000e+00  
12          1           0                     0   8.000000e+00  
13          0           0                     0   1.000000e+00

### Children 
Node 131031 has 0 children. Showing top-10 by descendant count

Empty DataFrame
Columns: []
Index: []

## Mutation path

site    node  descendants  inheritors inherited_state derived_state  \
23  25494  131031            1           1               G             C   
5    8358  131031            1           1               A             T   
10  17721  131031            1           1               G             T   
21  24745  135884          236         236               C             T   
29  29427  155262          262         262               G             A   
6    6539  155262          262         262               C             T   
11  20320  155262          262         262               C             T   
25  27739  141988          593         593               C             T   
12  20262  141988          586         586               A             G   
0    1267  141988          586         586               C             T   
8   12946  141988          586         586               T             C   
7    9891  134638         1797        1797               C             T   
9   11418  134638         1797        1797               T             C   
13  15451  142053        10584       10584               G             A   
14  16466  142053        10584       10567               C             T   
26  26767  149557        10685       10685               T             C   
27  27752  149557        10685       10684               C             T   
30  28461  149557        10685       10682               A             G   
22  24410  149557        10685       10685               G             A   
16  21618  149557        10685       10679               C             G   
18  22995  149557        10685       10685               C             A   
17  22917  149557        10685       10685               T             G   
1     210  149557        10685       10685               G             T   
19  23604  149557        10685       10681               C             G   
32  29402  118401        10810       10810               G             T   
24  25469  118401        10810       10810               C             T   
28  27638  118401        10810       10808               T             C   
31  28881  118401        10810       10810               G             T   
33  29742  118401        10810       10810               G             T   
2    5184    5299        10704        1917               C             T   
15  14408      58       150387      150284               C             T   
20  23403      15       150390      150389               A             G   
3     241      15       150389      150343               C             T   
4    3037      30       150394      150391               C             T   

    is_reversion  is_immediate_reversion  is_transition  is_transversion  \
23         False                   False          False             True   
5          False                   False          False             True   
10         False                   False          False             True   
21         False                   False           True            False   
29         False                   False           True            False   
6          False                   False           True            False   
11         False                   False           True            False   
25         False                   False           True            False   
12         False                   False           True            False   
0          False                   False           True            False   
8          False                   False           True            False   
7          False                   False           True            False   
9          False                   False           True            False   
13         False                   False           True            False   
14         False                   False           True            False   
26         False                   False           True            False   
27         False                   False           T

In [6]:
%%time
ts_path  ="../results/DELETE_v1-alpha4-mds1000/DELETE_v1-alpha4-mds1000_2021-03-22.ts"

import itertools
runs = []
for direction, num_mismatches in itertools.product(["forward", "reverse"], [3, 4, 5, 6, 7, 8, 9]):
    run = sc2ts.run_hmm(ds_path, ts_path, [focal_strain], direction=direction, num_mismatches=num_mismatches, 
                       deletions_as_missing=True, mismatch_threshold=40)
    runs.append(run)

len(runs)

CPU times: user 33min 55s, sys: 45.6 s, total: 34min 40s
Wall time: 35min 9s


16

In [7]:
data = []
for run in runs:
    r = run[0]
    data.append({
        "direction": r.direction,
        "num_mismatches": r.num_mismatches,
        "parents": len(r.match.path),
        "mutations": len(r.match.mutations)
    })
df = pd.DataFrame(data)
df

direction  num_mismatches  parents  mutations
0    forward               3        2         24
1    forward               4        2         24
2    forward               5        2         24
3    forward               6        1         29
4    forward               7        1         29
5    forward               8        1         29
6    forward               9        1         29
7    forward             100        1         29
8    reverse               3        2         24
9    reverse               4        2         24
10   reverse               5        2         24
11   reverse               6        1         29
12   reverse               7        1         29
13   reverse               8        1         29
14   reverse               9        1         29
15   reverse             100        1         29

For num_mismatches of < 6 we get a recombinant with 24 mutations and for greater >= 6 we have 29 mutations. How stable are these matches?

In [41]:
data = []
for run in runs:
    r = run[0]
    if len(r.match.path) > 1:
        data.append({
            "direction": r.direction,
            "num_mismatches": r.num_mismatches,
            "left_parent": r.match.path[0].parent,
            "right_parent": r.match.path[1].parent,
            "breakpoint": r.match.path[0].right,
            "mutations": [str(m) for m in r.match.mutations]
    })
dfr = pd.DataFrame(data)
dfr

direction  num_mismatches  left_parent  right_parent  breakpoint  \
0   forward               3         5299        118401       25469   
1   forward               4         5299        118401       25469   
2   forward               5         5299        118401       25469   
3   reverse               3        11294        119685       22023   
4   reverse               4        11294        119685       22023   
5   reverse               5        11294        119685       22023   

                                           mutations  
0  [G210T, C1267T, C6539T, A8358T, C9891T, T11418...  
1  [G210T, C1267T, C6539T, A8358T, C9891T, T11418...  
2  [G210T, C1267T, C6539T, A8358T, C9891T, T11418...  
3  [G210T, C1267T, C5184T, C6539T, A8358T, C9891T...  
4  [G210T, C1267T, C5184T, C6539T, A8358T, C9891T...  
5  [G210T, C1267T, C5184T, C6539T, A8358T, C9891T...

We do get a somewhat different solution on the reverse HMM pass, both in terms of breakpoint interval and the parent nodes. The breakpoint interval for the forward pass is 25277-25469, whereas the breakpoint on the reverse run is 22023 (we don't have an interval for that, currently).

In [48]:
data = []
for run in runs:
    r = run[0]
    if len(r.match.path) == 1:
        data.append({
            "direction": r.direction,
            "num_mismatches": r.num_mismatches,
            "parent": r.match.path[0].parent,
            "mutations": [str(m) for m in r.match.mutations]
    })
df_nor = pd.DataFrame(data)
df_nor

direction  num_mismatches  parent  \
0   forward               6    2910   
1   forward               7    2910   
2   forward               8    2910   
3   forward               9    2910   
4   forward             100    2910   
5   reverse               6    2910   
6   reverse               7    2910   
7   reverse               8    2910   
8   reverse               9    2910   
9   reverse             100    2910   

                                           mutations  
0  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
1  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
2  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
3  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
4  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
5  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
6  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
7  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
8  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...  
9  [G210T, C1267T, C5184T, A8358T, C9891T, T11418...

In [50]:
assert len(set([tuple(l) for l in df_nor.mutations])) == 1

The no recombinatin solutions are all the same.

Let's just filter down to the matches for num_mismatches = 5 and 6

In [84]:
hmm_runs = [r[0] for r in runs if r[0].num_mismatches in (5, 6)]
assert len(hmm_runs) == 4

In [90]:
fwd_re = hmm_runs[0]
rev_re = hmm_runs[2]
no_re = hmm_runs[1]
assert len(fwd_re.match.path) == 2
assert len(rev_re.match.path) == 2
assert len(no_re.match.path) == 1

In [91]:
fwd_re_mutations = set([str(m) for m in fwd_re.match.mutations])
rev_re_mutations = set([str(m) for m in rev_re.match.mutations])
no_re_mutations = set([str(m) for m in no_re.match.mutations])

In [92]:
common = sorted(fwd_re_mutations & rev_re_mutations & no_re_mutations, key=lambda m: int(m[1:-1]))
common, len(common)

(['G210T',
  'C1267T',
  'A8358T',
  'C9891T',
  'T11418C',
  'T12946C',
  'G15451A',
  'C16466T',
  'A20262G',
  'C20320T',
  'C21618G',
  'C22995A',
  'G24410A',
  'C24745T',
  'G25494C',
  'C27739T',
  'C27752T',
  'A28461G',
  'G29427A'],
 19)

There are 19 mutations common to all three solutions.

In [93]:
fwd_re_mutations - rev_re_mutations

{'C23604G', 'G17721T', 'T22917G', 'T26767C'}

In [94]:
rev_re_mutations - fwd_re_mutations

{'C23012G', 'C5184T', 'G26767C', 'T24775A'}

In [95]:
no_re_mutations - (fwd_re_mutations | rev_re_mutations)

{'C25469T', 'G28881T', 'G29402T', 'G29742T', 'T27638C'}

Here's the list of mutations for B.1.617 from the [designation issue](https://github.com/cov-lineages/pango-designation/issues/38) for reference

```
3457C>T	
4965C>T	
11201A>G	
17523G>T	
20396A>G	
21895T>C	 
21987G>A	 
22022G>A	 
22917T>G	 
23012G>C	 
23604C>G	 
24775A>T	 
25469C>T	 
27638T>C	 
28881G>T
``` 

In [100]:
muts_617 = [
    "C3457T",
    "C4965T",
    "A11201G",
    "G17523T",
    "A20396G",
    "T21895C",
    "G21987A",
    "G22022A",
    "T22917G",
    "G23012C",
    "C23604G",
    "A24775T",
    "C25469T",
    "T27638C",
    "G28881T"
]
no_re_mutations & set(muts_617)

{'C23604G', 'C25469T', 'G28881T', 'T22917G', 'T27638C'}

In [101]:
(no_re_mutations - (fwd_re_mutations | rev_re_mutations)) &  set(muts_617)

{'C25469T', 'G28881T', 'T27638C'}

In [106]:
site_id = np.searchsorted(ts.sites_position, [25469, 28881, 27638])
ti.sites_num_mutations[site_id]

array([76, 38,  3])

In [109]:
np.mean(ti.sites_num_mutations), np.median(ti.sites_num_mutations)

(8.373955642049458, 3.0)

So, of the mutations associated with the 617 designation only three of these are unique to the no-recombination case, and two of these occur at sites that are quite prone to recurrent mutaitions.

It's not clear at all really.

In [17]:
report(2910)

# Report for 2910

[{'sc2ts': {'date_added': '2020-03-27',
   'group_id': '4270e20b17ba02f3887bdb45b185d82e'}}]

### Node data 

node     flags              strain pango  parents  children  exact_matches  \
0  2910  ________            4270e20b              1         4              1   
1    58  _O______  2020-02-03:C14408T              1       302            257   

   descendants       date  mutations  reversions  immediate_reversions  \
0            4 2020-03-27          1           0                     0   
1       150387 2020-01-19          1           0                     0   

   child_mutations  child_reversions    role  branch_length  
0                1                 0   Child       0.000000  
1              484                 0  Parent      67.999996

### Edges

### Copying pattern

## Path to root

node     flags                             strain pango  parents  children  \
0  2910  ________                           4270e20b              1         4   
1    58  _O______                 2020-02-03:C14408T              1       302   
2    15  _O______  2020-01-28:A23403G, C3037T, C241T              1         3   
3    30  _O______                  2020-01-30:C3037T              1         3   
4     1  ______F_                    Wuhan/Hu-1/2019              1       130   

   exact_matches  descendants       date  mutations  reversions  \
0              1            4 2020-03-27          1           0   
1            257       150387 2020-01-19          1           0   
2              2       150390 2020-01-11          2           0   
3              0       150395 2020-01-03          1           0   
4             48       152905 2019-12-26          0           0   

   immediate_reversions  branch_length  
0                     0      67.999996  
1                     0       8.000000  
2                     0       8.000000  
3                     0       8.000000  
4                     0       1.000000

### Children 
Node 2910 has 4 children. Showing top-10 by descendant count

node     flags      strain pango  parents  children  exact_matches  \
0  2907  S_______  ERR6741493   B.1        1         0              0   
1  2908  S_______  ERR6741495   B.1        1         0              0   
2  2909  S_______  ERR6741496   B.1        1         0              0   
3  4938  S_______  ERR6739357   B.1        1         0              0   

   descendants       date  mutations  reversions  immediate_reversions  \
0            1 2020-03-27          0           0                     0   
1            1 2020-03-27          0           0                     0   
2            1 2020-03-27          0           0                     0   
3            1 2020-04-01          1           0                     0   

   child_mutations  child_reversions  new_muts  same_site_muts  branch_length  
0                0                 0         0               0       0.000004  
1                0                 0         0               0       0.000004  
2                0                 0         0               0       0.000004  
3                0                 0         1               0       5.000004

## Mutation path

site  node  descendants  inheritors inherited_state derived_state  \
2   6539  2910            4           4               C             T   
3  14408    58       150387      150284               C             T   
0    241    15       150389      150343               C             T   
4  23403    15       150390      150389               A             G   
1   3037    30       150394      150391               C             T   

   is_reversion  is_immediate_reversion  is_transition  is_transversion  \
2         False                   False           True            False   
3         False                   False           True            False   
0         False                   False           True            False   
4         False                   False           True            False   
1         False                   False           True            False   

   is_insertion  is_deletion  parent  num_parents        time      id  \
2         False        False      -1            0  451.000004   57266   
3         False        False      -1            0  519.000000  111632   
0         False        False      -1            0  527.000000    2464   
4         False        False      -1            0  527.000000  173869   
1         False        False      -1            0  535.000000   28335   

                                            metadata  
2  {'sc2ts': {'group_id': '4270e20b17ba02f3887bdb...  
3                     {'sc2ts': {'type': 'overlap'}}  
0                     {'sc2ts': {'type': 'overlap'}}  
4                     {'sc2ts': {'type': 'overlap'}}  
1                     {'sc2ts': {'type': 'overlap'}}

In [43]:
report(11294)

# Report for 11294

[{'sc2ts': {'date_added': '2020-04-30',
   'group_id': '83d634d4397afb943d9f8a4532aba256'}}]

### Node data 

node     flags    strain pango  parents  children  exact_matches  \
0  11294  ________  83d634d4              1         2              0   
1   6747  ________  5b3c8079              1        17             19   

   descendants       date  mutations  reversions  immediate_reversions  \
0            2 2020-04-30          1           0                     0   
1           34 2020-04-08          1           0                     0   

   child_mutations  child_reversions    role  branch_length  
0                0                 0   Child            0.0  
1               24                 0  Parent           22.0

### Edges

### Copying pattern

## Path to root

node     flags                                strain pango  parents  \
0  11294  ________                              83d634d4              1   
1   6747  ________                              5b3c8079              1   
2     97  _O______  2020-02-15:G28882A, G28881A, G28883C              1   
3     58  _O______                    2020-02-03:C14408T              1   
4     15  _O______     2020-01-28:A23403G, C3037T, C241T              1   
5     30  _O______                     2020-01-30:C3037T              1   
6      1  ______F_                       Wuhan/Hu-1/2019              1   

   children  exact_matches  descendants       date  mutations  reversions  \
0         2              0            2 2020-04-30          1           0   
1        17             19           34 2020-04-08          1           0   
2      1053           1257        85672 2020-01-30          3           0   
3       302            257       150387 2020-01-19          1           0   
4         3              2       150390 2020-01-11          2           0   
5         3              0       150395 2020-01-03          1           0   
6       130             48       152905 2019-12-26          0           0   

   immediate_reversions  branch_length  
0                     0      22.000000  
1                     0      68.999998  
2                     0      11.000000  
3                     0       8.000000  
4                     0       8.000000  
5                     0       8.000000  
6                     0       1.000000

### Children 
Node 11294 has 2 children. Showing top-10 by descendant count

node     flags      strain  pango  parents  children  exact_matches  \
0  11292  S_______  ERR4238637  B.1.1        1         0              0   
1  11293  S_______  ERR4248113  B.1.1        1         0              0   

   descendants       date  mutations  reversions  immediate_reversions  \
0            1 2020-04-30          0           0                     0   
1            1 2020-04-30          0           0                     0   

   child_mutations  child_reversions  new_muts  same_site_muts  branch_length  
0                0                 0         0               0       0.000002  
1                0                 0         0               0       0.000002

## Mutation path

site   node  descendants  inheritors inherited_state derived_state  \
2  17721  11294            2           2               G             T   
5  25728   6747           34          32               C             T   
6  28881     97        85671       85668               G             A   
7  28882     97        85671       85661               G             A   
8  28883     97        85671       85668               G             C   
3  14408     58       150387      150284               C             T   
0    241     15       150389      150343               C             T   
4  23403     15       150390      150389               A             G   
1   3037     30       150394      150391               C             T   

   is_reversion  is_immediate_reversion  is_transition  is_transversion  \
2         False                   False          False             True   
5         False                   False           True            False   
6         False                   False           True            False   
7         False                   False           True            False   
8         False                   False          False             True   
3         False                   False           True            False   
0         False                   False           True            False   
4         False                   False           True            False   
1         False                   False           True            False   

   is_insertion  is_deletion  parent  num_parents        time      id  \
2         False        False      -1            0  417.000002  130737   
5         False        False      -1            0  439.000002  197357   
6         False        False      -1            0  508.000000  234431   
7         False        False      -1            0  508.000000  234469   
8         False        False      -1            0  508.000000  234497   
3         False        False      -1            0  519.000000  111632   
0         False        False      -1            0  527.000000    2464   
4         False        False      -1            0  527.000000  173869   
1         False        False      -1            0  535.000000   28335   

                                            metadata  
2  {'sc2ts': {'group_id': '83d634d4397afb943d9f8a...  
5  {'sc2ts': {'group_id': '5b3c8079b73a1e7710e104...  
6                     {'sc2ts': {'type': 'overlap'}}  
7                     {'sc2ts': {'type': 'overlap'}}  
8                     {'sc2ts': {'type': 'overlap'}}  
3                     {'sc2ts': {'type': 'overlap'}}  
0                     {'sc2ts': {'type': 'overlap'}}  
4                     {'sc2ts': {'type': 'overlap'}}  
1                     {'sc2ts': {'type': 'overlap'}}

In [44]:
report(119685)

# Report for 119685

[{'sc2ts': {'date_added': '2021-03-04', 'mutations': ['G8137T', 'G25595T']}}]

### Node data 

node     flags                                strain pango  parents  \
0  119685  __P_____            2021-03-04:G8137T, G25595T              1   
1  118401  __P_____  2021-03-02:C20936T, C24863G, T27299C              1   

   children  exact_matches  descendants       date  mutations  reversions  \
0         3              0           85 2021-02-24          5           1   
1         3              0        10810 2021-01-13         16           0   

   immediate_reversions  child_mutations  child_reversions    role  \
0                     0                7                 0   Child   
1                     0                8                 1  Parent   

   branch_length  
0       0.000000  
1      42.492188

### Edges

### Copying pattern

## Path to root

node     flags                                strain pango  parents  \
0  119685  __P_____            2021-03-04:G8137T, G25595T              1   
1  118401  __P_____  2021-03-02:C20936T, C24863G, T27299C              1   
2  113172  _O______                    2021-02-22:C25276T              1   
3      58  _O______                    2020-02-03:C14408T              1   
4      15  _O______     2020-01-28:A23403G, C3037T, C241T              1   
5      30  _O______                     2020-01-30:C3037T              1   
6       1  ______F_                       Wuhan/Hu-1/2019              1   

   children  exact_matches  descendants       date  mutations  reversions  \
0         3              0           85 2021-02-24          5           1   
1         3              0        10810 2021-01-13         16           0   
2         2              0        10811 2020-04-04          1           0   
3       302            257       150387 2020-01-19          1           0   
4         3              2       150390 2020-01-11          2           0   
5         3              0       150395 2020-01-03          1           0   
6       130             48       152905 2019-12-26          0           0   

   immediate_reversions  branch_length  
0                     0      42.492188  
1                     0     283.937500  
2                     0      75.500000  
3                     0       8.000000  
4                     0       8.000000  
5                     0       8.000000  
6                     0       1.000000

### Children 
Node 119685 has 3 children. Showing top-10 by descendant count

node     flags              strain      pango  parents  children  \
0  121596  _O______  2021-03-07:C21846T                   1        17   
1  119168  ________            ce5821b7                   1         2   
2  118059  S______U          ERR5469699  B.1.617.1        1         0   

   exact_matches  descendants       date  mutations  reversions  \
0              0           64 2021-03-01          1           0   
1              0           20 2021-03-04          4           0   
2              0            1 2021-03-02          2           0   

   immediate_reversions  child_mutations  child_reversions  new_muts  \
0                     0               41                 0         1   
1                     0                1                 0         4   
2                     0                0                 0         2   

   same_site_muts  branch_length  
0               0       5.035156  
1               0       8.070312  
2               0       6.070312

## Mutation path

site    node  descendants  inheritors inherited_state derived_state  \
0     210  119685           85          85               G             T   
21  26767  119685           85          85               T             G   
18  25276  119685           85          85               T             C   
6   16852  119685           85          85               G             T   
7   20401  119685           85          85               T             G   
23  28881  118401        10810       10810               G             T   
22  27638  118401        10810       10808               T             C   
20  25469  118401        10810       10810               C             T   
17  24775  118401          125         125               A             T   
15  23604  118401          125         125               C             G   
14  23012  118401          125         125               G             C   
13  22917  118401          125         125               T             G   
12  22022  118401          125         124               G             A   
11  21895  118401          125         124               T             C   
9   20396  118401          125         125               A             G   
8   17523  118401          125         125               G             T   
5   11201  118401          125         125               A             G   
2    4965  118401          125         125               C             T   
1    3457  118401          125         125               C             T   
24  29402  118401        10810       10810               G             T   
25  29742  118401        10810       10810               G             T   
19  25276  113172          126           9               C             T   
10  14408      58       150387      150284               C             T   
16  23403      15       150390      150389               A             G   
3     241      15       150389      150343               C             T   
4    3037      30       150394      150391               C             T   

    is_reversion  is_immediate_reversion  is_transition  is_transversion  \
0          False                   False          False             True   
21         False                   False          False             True   
18          True                   False           True            False   
6          False                   False          False             True   
7          False                   False          False             True   
23         False                   False          False             True   
22         False                   False           True            False   
20         False                   False           True            False   
17         False                   False          False             True   
15         False                   False          False             True   
14         False                   False          False             True   
13         False                   False          False             True   
12         False                   False           True            False   
11         False                   False           True            False   
9          False                   False           True            False   
8          False                   False          False             True   
5          False                   False           True            False   
2          False                   False           True            False   
1          False                   False           True            False   
24         False                   False          False             True   
25         False                   False          False             True   
19         False                   False           True            False   
10         False                   False           True            False   
16         False                   False           True            False   
3          False                   False           T